src/main/java/com/google/maps/internal/GaePendingResult.java

```diff
@@ -0,0 +1,212 @@
+import com.google.maps.model.PriceLevel;
+import com.google.maps.model.TravelMode;
+import org.joda.time.DateTime;
+import org.joda.time.Instant;
+import org.joda.time.LocalTime;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.logging.Logger;
+
+/**
+ * A PendingResult backed by a HTTP call executed by Google App Engine URL Fetch capability,
+ * a deserialization step using Gson, and a retry policy.
+ *
+ * <p>{@code T} is the type of the result of this pending result, and {@code R} is the type of the
+ * request.
+ */
+public class GaePendingResult<T, R extends ApiResponse<T>>
+    implements PendingResult<T> {
+  private final HTTPRequest request;
+  private final URLFetchService client;
+  private final Class<R> responseClass;
+  private final FieldNamingPolicy fieldNamingPolicy;
+
+  private Callback<T> callback;
+  private long errorTimeOut;
+  private int retryCounter = 0;
+  private long cumulativeSleepTime = 0;
+  private Future<HTTPResponse> call;
+
+  private static final Logger LOG = Logger.getLogger(GaePendingResult.class.getName());
+  private static final List<Integer> RETRY_ERROR_CODES = Arrays.asList(500, 503, 504);
+
+  /**
+   * @param request           HTTP request to execute.
+   * @param client            The client used to execute the request.
+   * @param responseClass     Model class to unmarshal JSON body content.
+   * @param fieldNamingPolicy FieldNamingPolicy for unmarshaling JSON.
+   * @param errorTimeOut      Number of milliseconds to re-send erroring requests.
+   */
+  public GaePendingResult(HTTPRequest request, URLFetchService client, Class<R> responseClass,
+                          FieldNamingPolicy fieldNamingPolicy, long errorTimeOut) {
+    this.request = request;
+    this.client = client;
+    this.responseClass = responseClass;
+    this.fieldNamingPolicy = fieldNamingPolicy;
+    this.errorTimeOut = errorTimeOut;
+
+    this.call = client.fetchAsync(request);
+  }
+
+  @Override
+  public void setCallback(Callback<T> callback) {
+    throw new RuntimeException("setCallback not implemented for Google App Engine");
+  }
+
+  @Override
+  public T await() throws Exception {
+    HTTPResponse response = call.get();
+    if (response != null) {
+      return parseResponse(this, response);
+    } else {
+      throw new Exception(response.getResponseCode() + " " + new String(response.getContent()));
+    }
+  }
+
+  @Override
+  public T awaitIgnoreError() {
+    try {
+      return await();
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  @Override
+  public void cancel() {
+    call.cancel(true);
+  }
+
+
+  @SuppressWarnings("unchecked")
+  private T parseResponse(GaePendingResult<T, R> request, HTTPResponse response) throws Exception {
+    if (RETRY_ERROR_CODES.contains(response.getResponseCode()) && cumulativeSleepTime < errorTimeOut) {
```

Comment:
I tried to introduce an abstract base class, but I run into issues with type erasure. I could factor code into a delegate helper class or something. Thoughts?